File: src/test/java/net/flyingfishflash/loremlist/integration/domain/list/ListReadTest.java
package net.flyingfishflash.loremlist.integration.domain.list;

import static net.flyingfishflash.loremlist.core.response.advice.CoreExceptionHandler.VALIDATION_FAILURE_MESSAGE;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Stream;
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfProblem;
import net.flyingfishflash.loremlist.core.response.structure.DispositionOfSuccess;
import net.flyingfishflash.loremlist.domain.lrmlist.ListNotFoundException;
import net.flyingfishflash.loremlist.integration.domain.AbstractIntegrationTest;
import net.flyingfishflash.loremlist.integration.domain.TestData;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.http.HttpMethod;
import org.springframework.test.web.servlet.ResultMatcher;
import org.springframework.transaction.annotation.Transactional;

@Transactional
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ListReadTest extends AbstractIntegrationTest {

  private Map<UUID, LrmListCreateRequest> listIdMap;
  private int itemCount;

  @BeforeAll
  void setUp() throws Exception {
    purgeDomain();

    listIdMap = new LinkedHashMap<>();
    for (var request : TestData.LIST_CREATE_REQUESTS) {
      listIdMap.put(createAndVerifyList(request), request);
    }
    UUID firstListId = listIdMap.keySet().iterator().next();
    itemCount = 0;
    for (var request : TestData.ITEM_CREATE_REQUESTS_ALPHA) {
      createAndVerifyListItem(firstListId, request);
      itemCount++;
    }

    verifyContentValue("/lists/count", listIdMap.size());
  }

  @Test
  void allListsSucceedsWhenIncludeItemsIsTrue() throws Exception {
    verifyRequest(HttpMethod.GET, "/lists?includeItems=true")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .expectedSize(listIdMap.size())
        .additionalMatchers(
            jsonPath("$.content.length()").value(listIdMap.size()),
            jsonPath("$..items.length()").value(itemCount))
        .perform();
  }

  @Test
  void allListsSucceedsWhenIncludeItemsIsFalse() throws Exception {
    verifyRequest(HttpMethod.GET, "/lists?includeItems=false")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .expectedSize(listIdMap.size())
        .additionalMatchers(
            jsonPath("$.content.length()").value(listIdMap.size()),
            jsonPath("$..items.length()").value(0))
        .perform();
  }

  @Test
  void allListsSucceedsWhenIncludeItemsIsOmitted() throws Exception {
    verifyRequest(HttpMethod.GET, "/lists")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .expectedSize(listIdMap.size())
        .additionalMatchers(
            jsonPath("$.content.length()").value(listIdMap.size()),
            jsonPath("$..items.length()").value(0))
        .perform();
  }

  @Test
  void retrievingListFailsWhenInvalidUuidIsProvided() throws Exception {
    verifyRequest(HttpMethod.GET, "/lists/" + TestData.INVALID_UUIDS.get(0))
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isBadRequest())
        .additionalMatchers(
            jsonPath("$.message").value(VALIDATION_FAILURE_MESSAGE + " listId."),
            jsonPath("$.content.validationErrors.length()").value(1))
        .perform();
  }

  @Test
  void retrievingListFailsWhenNotFound() throws Exception {
    UUID listId = UUID.randomUUID();
    verifyRequest(HttpMethod.GET, "/lists/" + listId)
        .expectedDisposition(DispositionOfProblem.FAILURE)
        .statusMatcher(status().isNotFound())
        .additionalMatchers(
            jsonPath("$.content.title").value(ListNotFoundException.class.getSimpleName()),
            jsonPath("$.content.status").value("404"),
            jsonPath("$.content.supplemental.notFound").value(listId.toString()))
        .perform();
  }

  @ParameterizedTest(name = "succeeds when uuid is {0}")
  @MethodSource("listEntries")
  void retrievingListSucceedsWhenIncludeItemsIsTrue(UUID uuid, LrmListCreateRequest request)
      throws Exception {
    UUID firstListId = listIdMap.keySet().iterator().next();
    List<ResultMatcher> matchers =
        new ArrayList<>(
            List.of(
                jsonPath("$.content.id").value(uuid.toString()),
                jsonPath("$.content.name").value(request.name()),
                jsonPath("$.content.description").value(request.description()),
                jsonPath("$.content.public").value(request.isPublic())));
    matchers.add(jsonPath("$..items.length()").value(uuid.equals(firstListId) ? itemCount : 0));
    verifyRequest(HttpMethod.GET, "/lists/" + uuid + "?includeItems=true")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .additionalMatchers(matchers.toArray(new ResultMatcher[0]))
        .perform();
  }

  @ParameterizedTest(name = "succeeds when uuid is {0}")
  @MethodSource("listEntries")
  void retrievingListSucceedsWhenIncludeItemsIsFalse(UUID uuid, LrmListCreateRequest request)
      throws Exception {
    verifyRequest(HttpMethod.GET, "/lists/" + uuid + "?includeItems=false")
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .additionalMatchers(
            jsonPath("$.content.id").value(uuid.toString()),
            jsonPath("$.content.name").value(request.name()),
            jsonPath("$.content.description").value(request.description()),
            jsonPath("$.content.public").value(request.isPublic()))
        .perform();
  }

  @ParameterizedTest(name = "succeeds when uuid is {0}")
  @MethodSource("listEntries")
  void retrievingListSucceedsWhenIncludeItemsIsOmitted(UUID uuid, LrmListCreateRequest request)
      throws Exception {
    verifyRequest(HttpMethod.GET, "/lists/" + uuid)
        .expectedDisposition(DispositionOfSuccess.SUCCESS)
        .additionalMatchers(
            jsonPath("$.content.id").value(uuid.toString()),
            jsonPath("$.content.name").value(request.name()),
            jsonPath("$.content.description").value(request.description()),
            jsonPath("$.content.public").value(request.isPublic()))
        .perform();
  }

  Stream<Arguments> listEntries() {
    return listIdMap.entrySet().stream().map(e -> Arguments.of(e.getKey(), e.getValue()));
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/integration/domain/list/ListReadTest.kt
package net.flyingfishflash.loremlist.integration.domain.list

import io.kotest.datatest.withData
import net.flyingfishflash.loremlist.api.data.request.LrmItemCreateRequest
import net.flyingfishflash.loremlist.api.data.request.LrmListCreateRequest
import net.flyingfishflash.loremlist.core.response.advice.CoreExceptionHandler.Companion.VALIDATION_FAILURE_MESSAGE
import net.flyingfishflash.loremlist.domain.lrmlist.ListNotFoundException
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.invalidUuids
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.itemCreateRequestsAlpha
import net.flyingfishflash.loremlist.integration.domain.DomainFunctionTest.TestData.listCreateRequests
import org.springframework.http.HttpMethod
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.util.*

class ListReadTest(mockMvc: MockMvc) :
  DomainFunctionTest(mockMvc = mockMvc, body = {
    var listIdMap = mapOf<UUID, LrmListCreateRequest>()
    var itemIdMap = mapOf<UUID, LrmItemCreateRequest>()

    beforeSpec {
      purgeDomain()

      listIdMap = listCreateRequests.associateBy { request -> createAndVerifyList(request) }
      itemIdMap = itemCreateRequestsAlpha.associateBy { request ->
        createAndVerifyListItem(listId = listIdMap.keys.first(), itemRequest = request)
      }

      verifyContentValue("/lists/count", listIdMap.size)
    }

    describe("list retrieval") {
      context("retrieving all lists") {
        it("succeeds when ?includeItems=true") {
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists?includeItems=true",
            expectSuccess = true,
            expectedSize = listIdMap.size,
            additionalMatchers = arrayOf(
              jsonPath("$.content.length()").value(listIdMap.size),
              jsonPath("$..items.length()").value(itemIdMap.size),
            ),
          )
        }

        it("succeeds when ?includeItems=false") {
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists?includeItems=false",
            expectSuccess = true,
            expectedSize = listIdMap.size,
            additionalMatchers = arrayOf(
              jsonPath("$.content.length()").value(listIdMap.size),
              jsonPath("$..items.length()").value(0),
            ),
          )
        }

        it("succeeds when ?includeItems is omitted (default behavior)") {
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists",
            expectSuccess = true,
            expectedSize = listIdMap.size,
            additionalMatchers = arrayOf(
              jsonPath("$.content.length()").value(listIdMap.size),
              jsonPath("$..items.length()").value(0),
            ),
          )
        }
      }

      context("retrieving a list") {
        it("fails when list is not found") {
          val listId = UUID.randomUUID()
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists/$listId",
            expectSuccess = false,
            statusMatcher = status().isNotFound(),
            additionalMatchers = arrayOf(
              jsonPath("$.content.title").value(ListNotFoundException::class.java.simpleName),
              jsonPath("$.content.status").value("404"),
              jsonPath("$.content.supplemental.notFound").value(listId.toString()),
            ),
          )
        }
      }

      context("retrieving a list when ?includeItems is true") {
        withData(
          nameFn = { uuid -> "succeeds when uuid is $uuid" },
          listIdMap.entries.toList(),
        ) { (uuid, request) ->
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists/$uuid?includeItems=true",
            expectSuccess = true,
            additionalMatchers = arrayOf(
              jsonPath("$.content.id").value(uuid.toString()),
              jsonPath("$.content.name").value(request.name),
              jsonPath("$.content.description").value(request.description),
              jsonPath("$.content.public").value(request.public),
              if (uuid == listIdMap.keys.first()) {
                jsonPath("$..items.length()").value(itemIdMap.size)
              } else {
                jsonPath("$..items.length()").value(0)
              },

            ),
          )
        }
      }

      context("retrieving a list when ?includeItems is false") {
        withData(
          nameFn = { uuid -> "succeeds when uuid is $uuid" },
          listIdMap.entries.toList(),
        ) { (uuid, request) ->
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists/$uuid?includeItems=false",
            expectSuccess = true,
            additionalMatchers = arrayOf(
              jsonPath("$.content.id").value(uuid.toString()),
              jsonPath("$.content.name").value(request.name),
              jsonPath("$.content.description").value(request.description),
              jsonPath("$.content.public").value(request.public),
            ),
          )
        }
      }

      context("retrieving a list when ?includeItems is omitted (default behavior)") {
        withData(
          nameFn = { uuid -> "succeeds when uuid is $uuid" },
          listIdMap.entries.toList(),
        ) { (uuid, request) ->
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists/$uuid",
            expectSuccess = true,
            additionalMatchers = arrayOf(
              jsonPath("$.content.id").value(uuid.toString()),
              jsonPath("$.content.name").value(request.name),
              jsonPath("$.content.description").value(request.description),
              jsonPath("$.content.public").value(request.public),
            ),
          )
        }
      }

      context("path variable validation") {
        it("fails when invalid uuid is provided for list id") {
          performRequestAndVerifyResponse(
            method = HttpMethod.GET,
            instance = "/lists/${invalidUuids[0]}",
            expectSuccess = false,
            statusMatcher = status().isBadRequest(),
            additionalMatchers = arrayOf(
              jsonPath("$.message").value("$VALIDATION_FAILURE_MESSAGE listId."),
              jsonPath("$.content.validationErrors.length()").value(1),
            ),
          )
        }
      }
    }
  })
